### PR TITLE
feat(agent): add script data dir for binaries and files

### LINKED ISSUE #11131: Allow the coder agent to tell modules where they can place data, binaries, etc
This is an alternate path to achieving `PATH` manipulations as discussed in: https://github.com/coder/coder/issues/10166

The premise is, let the agent decide where a module can store its files (binaries, data, settings, etc).

Let's say the agent discerns that `$HOME` is a good place to put files, it can create the appropriate folders/permissions and run module script(s) with the following env variables:

```shell
CODER_MODULE_DATA_PATH=~/.local/share/coder_module/${MODULE_ID}
CODER_MODULE_BIN_PATH=~/.local/share/coder_module/${MODULE_ID}/bin
```

Or let's say `$HOME` isn't writable, we can usually default to the `agent` directory often located in `/tmp` (granted, this is ephemeral storage, so a warning may be prudent).

```shell
CODER_MODULE_DATA_PATH=/tmp/coder.xxcxcv/module/${MODULE_ID}
CODER_MODULE_BIN_PATH=/tmp/coder.xxcxcv/module/${MODULE_ID}/bin
```

As long as we can create an association between coder module (name/id) and whatever `coder_script` it has introduced, we should be able to do this fairly easily in the agent.

We can either use a per-module `/bin` or a combine bin dir, e.g. 

```shell
CODER_MODULE_BIN_PATH=~/.local/share/coder_module/bin
```

Lastly, the agent would make sure to prepend (or append) the module bin dir(s) to `PATH` for use in scripts, SSH sessions, etc.

### PR DESCRIPTION
**Pull request by mafredri**
The agent is extended with a `--script-data-dir` flag, defaulting to the
OS temp dir. This dir is used for storing `coder-script-data/bin` and
`coder-script/[script uuid]`. The former is a place for all scripts to
place executable binaries that will be available by other scripts, SSH
sessions, etc. The latter is a place for the script to store files.

Since we default to OS temp dir, files are ephemeral by default. In the
future, we may consider adding new env vars or changing the default
storage location. Workspace startup speed could potentially benefit from
scripts being able to skip steps that require downloading software. We
may also extend this with more env variables (e.g. persistent storage in
HOME).

Fixes #11131

This can be feature detected by checking if `$CODER_SCRIPT_DATA_DIR` and/or `$CODER_SCRIPT_BIN_DIR` is set.

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -953,6 +963,13 @@ func (a *agent) updateCommandEnv(current []string) (updated []string, err error)
 		envs[k] = v
 	}
 
+	// Prepend the agent script bin directory to the PATH
+	// (this is where Coder modules place their binaries).
+	if _, ok := envs["PATH"]; !ok {
+		envs["PATH"] = os.Getenv("PATH")
+	}
+	envs["PATH"] = a.scriptRunner.ScriptBinDir() + ":" + envs["PATH"]
```
Comment (johnstcn) [inline agent/agent.go:971]: I think in Windows you need to use `;` to separate path elements.

File: agent/agentscripts/agentscripts.go [hunk 2]
```diff
@@ -238,6 +263,13 @@ func (r *Runner) run(ctx context.Context, script codersdk.WorkspaceAgentScript)
 	cmd.WaitDelay = 10 * time.Second
 	cmd.Cancel = cmdCancel(cmd)
 
+	// Expose env vars that can be used in the script for storing data
+	// and binaries. In the future, we may want to expose more env vars
+	// for the script to use, like CODER_SCRIPT_DATA_DIR for persistent
+	// storage.
+	cmd.Env = append(cmd.Env, "CODER_SCRIPT_DATA_DIR="+scriptDataDir)
+	cmd.Env = append(cmd.Env, "CODER_SCRIPT_BIN_DIR="+r.ScriptBinDir())
```
Comment (johnstcn) [inline agent/agentscripts/agentscripts.go:271]: Will bin dir always be a subpath of data dir?

**Review by mtojek — APPROVED** [review]
Nothing blocking!

File: agent/agent.go (hunk 1 above)
Comment (mafredri) [reply agent/agent.go]: Thanks for noticing! @mtojek no need, we can simply use `filepath.ListSeparator`.

File: agent/agentscripts/agentscripts.go (hunk 2 above)
Comment (mafredri) [reply agent/agentscripts/agentscripts.go:271]: @johnstcn yes, that's the intent. All script-data (inc. bin) will be contained in a single path for tidyness.

Comment (johnstcn) [reply agent/agentscripts/agentscripts.go:271]: Do we need both environment variables then?

Comment (mafredri) [reply agent/agentscripts/agentscripts.go:271]: The second is not static, it's `datadir/uuid`.

Comment (johnstcn) [reply agent/agentscripts/agentscripts.go:271]: Ah, that makes much more sense then 👍

**Review by johnstcn — APPROVED** [review]

File: cli/agent.go [hunk 3]
```diff
@@ -339,6 +341,13 @@ func (r *RootCmd) workspaceAgent() *clibase.Cmd {
 			Env:         "CODER_AGENT_LOG_DIR",
 			Value:       clibase.StringOf(&logDir),
 		},
+		{
+			Flag:        "script-data-dir",
+			Default:     os.TempDir(),
+			Description: "Specify the location for storing script data.",
+			Env:         "CODER_AGENT_SCRIPT_DIR",
```
Comment (mafredri) [inline cli/agent.go:348]:
```suggestion
			Env:         "CODER_AGENT_SCRIPT_DATA_DIR",
```

File: cli/testdata/coder_agent_--help.golden [hunk 4]
```diff
@@ -33,6 +33,9 @@ OPTIONS:
       --prometheus-address string, $CODER_AGENT_PROMETHEUS_ADDRESS (default: 127.0.0.1:2112)
           The bind address to serve Prometheus metrics.
 
+      --script-data-dir string, $CODER_AGENT_SCRIPT_DIR (default: /tmp)
```
Comment (mafredri) [inline cli/testdata/coder_agent_--help.golden:36]:
```suggestion
      --script-data-dir string, $CODER_AGENT_SCRIPT_DATA_DIR (default: /tmp)
```

File: agent/agent_test.go [hunk 5]
```diff
@@ -294,7 +300,10 @@ func TestAgent_Session_EnvironmentVariables(t *testing.T) {
 		},
 	}
 	banner := codersdk.ServiceBannerConfig{}
+	fs := afero.NewMemMapFs()
```
Comment (mafredri) [inline agent/agent_test.go:303]:
```suggestion
```

File: agent/agent_test.go [hunk 6]
```diff
@@ -294,7 +300,10 @@ func TestAgent_Session_EnvironmentVariables(t *testing.T) {
...
 	session := setupSSHSession(t, manifest, banner, nil, func(_ *agenttest.Client, opts *agent.Options) {
+		opts.Filesystem = fs
```
Comment (mafredri) [inline agent/agent_test.go:305]:
```suggestion
```

**Pull request merged**